### PR TITLE
Generate remote binary

### PR DESCRIPTION
**Pull request by AndrienkoAleksandr**
### What does this PR do?
Add image with remote-plugin-endpoint binary

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/13387

### CODE REVIEW

**Comment by dmytro-ndp** [comment]
crw-ci-test

File: dockerfiles/theia-endpoint-runtime/Dockerfile [hunk 1]
```diff
@@ -8,54 +8,26 @@
...
-        if [ "${GITHUB_LIMIT}" -lt 10 ]; then \
-            printf "\033[0;31m\n\n\nRate limit on https://api.github.com is reached so in order to build this image, "; \
-            printf "the build argument GITHUB_TOKEN needs to be provided so build will not fail.\n\n\n\033[0m"; \
-            exit 1; \
-        else \
-            echo "GITHUB_TOKEN variable is not set but https://api.github.com rate limit has enough slots"; \
-        fi \
-    fi
+# Light image without node. We include remote binary to this image.
+FROM alpine
```
Comment (benoitf) [inline dockerfiles/theia-endpoint-runtime/Dockerfile]: hello, please add the version for the image (to be sure to not get unwanted) updates

File: dockerfiles/theia-endpoint-runtime/Dockerfile [hunk 2]
```diff
@@ -8,54 +8,26 @@
...
 
-#invalidate cashe
-ADD https://${GITHUB_TOKEN}:x-oauth-basic@api.github.com/repos/theia-ide/theia/git/refs/head /tmp/branch_info.json
-ADD https://${GITHUB_TOKEN}:x-oauth-basic@api.github.com/repos/eclipse/che-theia/git/refs/head /tmp/branch_info.json
+RUN mkdir /home/theia
```
Comment (benoitf) [inline dockerfiles/theia-endpoint-runtime/Dockerfile]: Do we need this docker image if we do injection of the binary ?

or it's to have the binary for injection that we're making this image ?

File: dockerfiles/theia-endpoint-runtime/etc/entrypoint.sh [hunk 3]
```diff
@@ -47,12 +47,11 @@ set -e
 
 # setup handlers
 # on callback, kill the last background process, which is `tail -f /dev/null` and execute the specified handler
-trap 'responsible_shutdown' SIGHUP SIGTERM SIGINT
-
-cd ${HOME}
+trap 'responsible_shutdown' HUP TERM INT
```
Comment (benoitf) [inline dockerfiles/theia-endpoint-runtime/etc/entrypoint.sh]: is there a reason to change the traps ?

File: extensions/eclipse-che-theia-plugin-ext/src/node/che-plugin-api-provider.ts [hunk 4]
```diff
@@ -22,7 +21,7 @@ export class ChePluginApiProvider implements ExtPluginApiProvider {
                 initFunction: 'initializeApi',
                 initVariable: 'che_api_provider'
             },
-            backendInitPath: path.join(__dirname, '../plugin/node/che-api-node-provider.js')
+            backendInitPath: '@eclipse-che/theia-plugin-ext/lib/plugin/node/che-api-node-provider.js'
```
Comment (benoitf) [inline extensions/eclipse-che-theia-plugin-ext/src/node/che-plugin-api-provider.ts]: maybe this change and the next one could be merged in another PR

**Comment by dmytro-ndp** [comment]
crw-ci-test

File: dockerfiles/theia-endpoint-runtime/Dockerfile (hunk 2 above)
Comment (AndrienkoAleksandr) [reply dockerfiles/theia-endpoint-runtime/Dockerfile]: This image will be init container for Che theia editor plugin.

File: dockerfiles/theia-endpoint-runtime/etc/entrypoint.sh (hunk 3 above)
Comment (AndrienkoAleksandr) [reply dockerfiles/theia-endpoint-runtime/etc/entrypoint.sh]: Yes, of course, SIGHUP SIGTERM SIGINT doesn't work on fedora.

**Comment by AndrienkoAleksandr** [comment]
ci-build-check

**Comment by AndrienkoAleksandr** [comment]
ci-build

**Comment by AndrienkoAleksandr** [comment]
@benoitf @evidolob @mmorhun Could you review, please?

File: dockerfiles/theia-endpoint-runtime-binary/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+USER root
+
+RUN cd /home/theia && \
+    yarn global add nexe@3.3.2 && nexe -v && \
+    # Build remote binary with node runtime 10.14.2 and che-theia node dependencies. nexe icludes to the binary only
+    # necessary dependencies.
+    nexe node_modules/@eclipse-che/theia-remote/lib/node/plugin-remote.js -t alpine-x64-10.16.0 -o /plugin-remote-endpoint
+
+# Light image without node. We include remote binary to this image.
+FROM alpine:3.10.2
```
Comment (benoitf) [inline dockerfiles/theia-endpoint-runtime-binary/Dockerfile]: what is the difference of using FROM ALPINE:3.10.2 and using FROM scratch if we don't install anything and that the binary is self-contained ? it's due to musl linking I suppose ?

File: dockerfiles/theia-endpoint-runtime-binary/build.sh [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+
+base_dir=$(cd "$(dirname "$0")"; pwd)
+. "${base_dir}"/../build.include
+
+init --name:theia-endpoint-runtime-binary "$@"
+build
```
Comment (benoitf) [inline dockerfiles/theia-endpoint-runtime-binary/build.sh]: very minor: missing EOF

File: dockerfiles/theia-endpoint-runtime-binary/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+
+FROM ${BUILD_ORGANIZATION}/${BUILD_PREFIX}-theia:${BUILD_TAG} as builder
+
+USER root
+
+RUN cd /home/theia && \
```
Comment (benoitf) [inline dockerfiles/theia-endpoint-runtime-binary/Dockerfile]: it seems usually linters want that we use WORKDIR /home/theia before but I haven't run linter there so maybe it's fine

File: dockerfiles/theia-endpoint-runtime-binary/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+
+FROM ${BUILD_ORGANIZATION}/${BUILD_PREFIX}-theia:${BUILD_TAG} as builder
+
+USER root
```
Comment (benoitf) [inline dockerfiles/theia-endpoint-runtime-binary/Dockerfile]: do we need to be root ?
AFAIK we can install stuff globally in theia images ?

File: dockerfiles/theia-endpoint-runtime-binary/Dockerfile (hunk 5 above)
Comment (AndrienkoAleksandr) [reply dockerfiles/theia-endpoint-runtime-binary/Dockerfile]: Sorry, @benoitf, Seems, scratch has only latest tag. And this image has not `cp` and  has not shell(even without sh).  Dockerfile `RUN` section doesn't work without sh, so I can not install cp. And I think scratch has not any package manager, because it's empty image, so I don't know how to install cp or so on.

Comment (benoitf) [reply dockerfiles/theia-endpoint-runtime-binary/Dockerfile]: ok I see.

I might see another problem.
We generate nexe for alpine but if the runtime is not alpine, what is the behavior ?
like using debian, centos, fedora, etc images ?

as we'll copy the bootstrap to another container ?

Comment (AndrienkoAleksandr) [reply dockerfiles/theia-endpoint-runtime-binary/Dockerfile]: @benoitf I tested with centos, fedora debian and it is works. List tested images was pretty big. You can see this list https://github.com/eclipse/che-theia/pull/267/commits/99ae354ed9586918703026eec7f0c8ce18d01267#diff-f7eaf7e082319e42b2dfd26a86ca0736R80 search by from, I commented code, but each image was tested.

Comment (benoitf) [reply dockerfiles/theia-endpoint-runtime-binary/Dockerfile]: ok, still strange but if it works.

**Comment by benoitf** [comment]
hello, how to test ?

**Comment by AndrienkoAleksandr** [comment]
> @benoitf  hello, how to test ?

**All stuff I tested on the minishift.**

### Build che-theia and theia-endpoint-runtime-binary images

@You need to build images che-theia, and theia-endpoint-runtime-binary images, so from my branch: 

```
./build.sh --pr
```

You can retag and push images to the dockerhub.

### Use image with upgraded che-plugin-broker

For integration image theia-endpoint-runtime-binary like init-container for che-theia, you need to use upgraded che-plugin-broker: https://github.com/eclipse/che-plugin-broker/pull/75 . You can use mine image: `aandrienko/che-unified-plugin-broker:latest`, or build own using script(but patched to use your dockerhub account): https://github.com/eclipse/che-plugin-broker/tree/startRemoteBinaryWithOverrideEntryPoint/.ci/build.sh

To activate this plugin-broker image in the che, you need to open deployment che-server and apply env variable:

CHE_WORKSPACE_PLUGIN__BROKER_UNIFIED_IMAGE aandrienko/che-unified-plugin-broker:latest

Redeploy che.

### Test devfile

And you can use some test devfile with your che-theia and theia-endpoint-runtime-binary images. Example of mine devfile: 

https://gist.github.com/AndrienkoAleksandr/ec1efffcbb4b25d52c22029bb7b6ac69

### Interesting part of the test devfile.

We are using our builded image with che-theia: 

https://github.com/AndrienkoAleksandr/che-plugin-registry/blob/master/v3/plugins/eclipse/che-theia/remote2/meta.yaml#L63

Init container section here with theia-endpoint-runtime-binary:

https://github.com/AndrienkoAleksandr/che-plugin-registry/blob/master/v3/plugins/eclipse/che-theia/remote2/meta.yaml#L51

### Sidecar image for typescript lsp

And we need image for sidecar container, here is I used fedora:30:

[DockerFile](https://raw.githubusercontent.com/AndrienkoAleksandr/che-plugin-registry/master/v3/plugins/che-incubator/typescript/fedora/Dockerfile)
[Fedora with node image build Script(should be used for typescript lsp)](https://raw.githubusercontent.com/AndrienkoAleksandr/che-plugin-registry/master/v3/plugins/che-incubator/typescript/fedora/buildTestPluginImage.sh)
[meta.yaml](https://raw.githubusercontent.com/AndrienkoAleksandr/che-plugin-registry/master/v3/plugins/che-incubator/typescript/fedora/meta.yaml)

In the devfile sidecar plugin here:
https://gist.github.com/AndrienkoAleksandr/ec1efffcbb4b25d52c22029bb7b6ac69#file-gistfile1-txt-L16

**Comment by AndrienkoAleksandr** [comment]
@l0rd @benoitf Is it OK to merge it? It's planned for 7.3 and rework sidecar images depends on it.

**Comment by benoitf** [comment]
yes
it's not harmful to merge it BTW

**Comment by AndrienkoAleksandr** [comment]
ci-build

**Comment by AndrienkoAleksandr** [comment]
crw-ci-test

**Comment by musienko-maxim** [comment]
crw-ci-test

**Comment by musienko-maxim** [comment]
The PR was checked manually. All steps in the HappyPath test were passed.
